File: ST_engine/src/main/game_manager/level/light.cpp
/* This file is part of the "ST" project.
 * You may use, distribute or modify this code under the terms
 * of the GNU General Public License version 2.
 * See LICENCE.txt in the root directory of the project.
 *
 * Author: Maxim Atanasov
 * E-mail: maxim.atanasov@protonmail.com
 */

#include <game_manager/level/light.hpp>

/**
 * The constructor for a light object.
 * @param origin_x The X position of the light.
 * @param origin_y The Y position of the light.
 * @param radius The radius of the light.
 * @param intensity The intensity of the light.
 * @param brightness The brightness of the light.
 */
ST::light::light(int32_t origin_x, int32_t origin_y, uint16_t radius, uint16_t intensity, uint16_t brightness){
    this->origin_x = origin_x;
    this->origin_y = origin_y;
    this->radius = radius;
    this->intensity = intensity;
    this->brightness = brightness;
}

File: ST_engine/gamedata/lua/objects/light.lua
-- This file is part of the "ST" project.
-- You may use, distribute or modify this code under the terms
-- of the GNU General Public License version 2.
-- See LICENCE.txt in the root directory of the project.
--
-- Author: Maxim Atanasov
-- E-mail: maxim.atanasov@protonmail.com

--LIGHTING
function newLightID()
    local temp = currentLightID
    currentLightID = currentLightID + 1;
    return temp
end

light = {
    ID;
    origin_x;
    origin_y;
    radius = 0;
    intensity = 0;
    brightness = 0;
}

function newLight(self, origin_x, origin_y, radius, intensity, brightness)
    local o = {}
    setmetatable(o, self)
    self.__index = self
    if(origin_x ~= nil and origin_y ~= nil) then
        o.ID = newLightID()
        createLight(origin_x, origin_y, radius, intensity, brightness)
    end
    return o
end


--Creates a new light object and returns it
function light:new(origin_x, origin_y, radius, intensity, brightness)
    return newLight(self, origin_x, origin_y, radius, intensity, brightness)
end

--Simple getters and setters

function light:setOriginX(origin_x)
    setLightOriginX(self.ID, origin_x)
end

function light:getOriginX()
    return getLightOriginX(self.ID)
end

function light:setOriginY(origin_y)
    setLightOriginY(self.ID, origin_y)
end

function light:getOriginY()
    return getLightOriginY(self.ID)
end

function light:setBrightness(brightness)
    setLightBrightness(self.ID, brightness)
end

function light:getBrightness()
    return getLightBrightness(self.ID)
end

function light:setRadius(radius)
    setLightRadius(self.ID, radius)
end

function light:getRadius()
    return getLightRadius(self.ID)
end

function light:setIntensity(intensity)
    setLightIntensity(self.ID, intensity)
end

function light:getIntensity()
    return getLightIntensity(self.ID);
end

function light:setStatic(arg)
    setLightStatic(self.ID, arg)
end

function light:isStatic()
    return isLightStatic(self.ID)
end